File: player/shared/src/commonMain/kotlin/com/spela/player/domain/model/PlaySemantics.kt
package com.spela.player.domain.model

/**
 * Three-state taxonomy for the game-detail hero's primary action
 * label. Captures the difference between "user has played this and
 * the engine will pick up where they left off" vs. "user has played
 * this but launching takes them to the title screen and they have to
 * navigate to the engine's own save menu."
 *
 * Resolved from `(hasSession, consoleSaveStateSupport, effectiveSaveStateChoice)`
 * — see [resolvePlaySemantics]. The view-model maps this enum to the
 * label string the hero renders. See #884.
 */
enum class PlaySemantics {
    /** No session for this game yet. Label: "Play". */
    NoSession,

    /**
     * Session exists AND save state will auto-load on launch — the
     * console supports save states *and* the user hasn't opted out
     * via the per-console / per-game policy from #804. The user lands
     * mid-frame where they left off. Label: "Resume".
     */
    ResumesFromSaveState,

    /**
     * Session exists AND save state will NOT auto-load on launch —
     * the console doesn't support save states (e.g. ScummVM) or the
     * user opted out for this console / game. Launching takes the
     * user to the title screen; in-engine save menus are on them.
     * Label: "Continue" — honest about the difference, vs. "Resume"
     * which would over-promise.
     */
    LaunchesFresh,
}

/**
 * Pure resolver. The four open inputs come from these existing models:
 *
 *  - [hasSession] — `state.sessions.isNotEmpty()` on game-detail.
 *  - [consoleSaveStateSupport] — `Console.saveStateSupport`. Hardcoded
 *    server-side for cores that genuinely cannot snapshot RAM (ScummVM,
 *    demo cores).
 *  - [effectiveChoice] — already resolved upstream by
 *    [effectiveSaveStateChoice]; the resolver here only cares whether
 *    it's `Disabled` (no auto-load) or anything else (auto-load on).
 *    `AskOnce` counts as auto-load — the in-game flow treats it as
 *    Enabled until the prompt resolves to a deliberate choice.
 *
 * Pure function — no IO, no platform calls. Trivial to unit test, and
 * cheap enough to recompute every recomposition.
 */
fun resolvePlaySemantics(
    hasSession: Boolean,
    consoleSaveStateSupport: Boolean,
    effectiveChoice: SaveStateChoice,
): PlaySemantics {
    if (!hasSession) return PlaySemantics.NoSession
    val willAutoLoad = consoleSaveStateSupport && effectiveChoice != SaveStateChoice.Disabled
    return if (willAutoLoad) PlaySemantics.ResumesFromSaveState else PlaySemantics.LaunchesFresh
}

File: web/src/lib/play-semantics.ts
/**
 * Mirror of the Kotlin `PlaySemantics` enum + resolver in
 * `player/shared/src/commonMain/kotlin/com/spela/player/domain/model/PlaySemantics.kt`.
 *
 * Three states drive the game-detail hero's primary action label so
 * ScummVM / demo-core / save-state-disabled launches don't promise
 * "Resume" when the engine is going to land on the title screen.
 *
 * Resolver inputs match the Kotlin side exactly; keep the two in sync
 * whenever either side changes — there's no codegen across, just a
 * thin parallel implementation. See #884 / #900.
 */
export type PlaySemantics =
  | "no-session"
  | "resumes-from-save-state"
  | "launches-fresh";

export type SaveStateChoice = "enabled" | "disabled" | "ask-once";

/**
 * Resolves the effective save-state policy for a console + game.
 * Mirrors the Kotlin `effectiveSaveStateChoice` resolver.
 *
 * Precedence: per-game override → per-console override → tier default.
 * For the hero label resolver we don't need the tier branch — any
 * SaveStateChoice that isn't `"disabled"` counts as auto-load.
 */
export function effectiveSaveStateChoice(
  consoleAbbr: string,
  consoleSaveStatePolicies: Readonly<Record<string, SaveStateChoice>>,
  gameOverride: SaveStateChoice | null | undefined,
): SaveStateChoice {
  if (gameOverride) return gameOverride;
  const key = consoleAbbr.toLowerCase();
  return consoleSaveStatePolicies[key] ?? "enabled";
}

export interface ResolvePlaySemanticsInput {
  hasSession: boolean;
  consoleSaveStateSupport: boolean;
  effectiveChoice: SaveStateChoice;
}

export function resolvePlaySemantics({
  hasSession,
  consoleSaveStateSupport,
  effectiveChoice,
}: ResolvePlaySemanticsInput): PlaySemantics {
  if (!hasSession) return "no-session";
  const willAutoLoad =
    consoleSaveStateSupport && effectiveChoice !== "disabled";
  return willAutoLoad ? "resumes-from-save-state" : "launches-fresh";
}

/**
 * The user-facing label for each state. Matches the player app
 * exactly (#900 — "New game" / "Resume" / "Continue").
 */
export function playSemanticsLabel(s: PlaySemantics): string {
  switch (s) {
    case "no-session":
      return "New game";
    case "resumes-from-save-state":
      return "Resume";
    case "launches-fresh":
      return "Continue";
  }
}
